README.md
# NAME

Bot::ChatBots::Telegram - Telegram adapter for Bot::ChatBots

# VERSION

This document describes Bot::ChatBots::Telegram version {{\[ version \]}}.

<div>
    <a href="https://travis-ci.org/polettix/Bot-ChatBots-Telegram">
    <img alt="Build Status" src="https://travis-ci.org/polettix/Bot-ChatBots-Telegram.svg?branch=master">
    </a>
    <a href="https://www.perl.org/">
    <img alt="Perl Version" src="https://img.shields.io/badge/perl-5.10+-brightgreen.svg">
    </a>
    <a href="https://badge.fury.io/pl/Bot-ChatBots-Telegram">
    <img alt="Current CPAN version" src="https://badge.fury.io/pl/Bot-ChatBots-Telegram.svg">
    </a>
    <a href="http://cpants.cpanauthors.org/dist/Bot-ChatBots-Telegram">
    <img alt="Kwalitee" src="http://cpants.cpanauthors.org/dist/Bot-ChatBots-Telegram.png">
    </a>
    <a href="http://www.cpantesters.org/distro/B/Bot-ChatBots-Telegram.html?distmat=1">
    <img alt="CPAN Testers" src="https://img.shields.io/badge/cpan-testers-blue.svg">
    </a>
    <a href="http://matrix.cpantesters.org/?dist=Bot-ChatBots-Telegram">
    <img alt="CPAN Testers Matrix" src="https://img.shields.io/badge/matrix-@testers-blue.svg">
    </a>
</div>

# SYNOPSIS

    # A minimal Telegram Bot using WebHooks
    use Log::Any qw< $log >;
    use Log::Any::Adapter;
    use Mojolicious::Lite;
    Log::Any::Adapter->set(MojoLog => logger => app->log);
    plugin 'Bot::ChatBots::Telegram' => instances => [
       [
          'WebHook',
          processor  => \&processor,
          register   => 1,
          token      => $ENV{TOKEN},
          unregister => 1,
          url        => 'https://example.com:8443/mybot',
       ],
       # more can follow here...
    ];
    app->start;
    sub processor {
       my $record = shift;
       # do whatever you want with $record, e.g. set a quick response
       $record->{send_response} = 'your thoughs are important for us!';
       return $record;
    }

    # You can also add Bot::ChatBots::Telegram::LongPoll sources if you want

# DESCRIPTION

This module allows you to to define [Bot::ChatBots](https://metacpan.org/pod/Bot::ChatBots) for
[Telegram](https://telegram.org/), a messaging application. For an
introduction and tutorial, see
[Bot::ChatBots::Telegram::Guide::Tutorial](https://metacpan.org/pod/Bot::ChatBots::Telegram::Guide::Tutorial).

Strictly speaking, this _module_ is a [Mojolicious](https://metacpan.org/pod/Mojolicious) plugin that allows
you to load and use [Bot::ChatBots::Telegram::WebHook](https://metacpan.org/pod/Bot::ChatBots::Telegram::WebHook) (see ["SYNOPSIS"](#synopsis)
for a quick example). On the other hand, the _distribution_ also contains
other modules that allow you to interact with Telegram, e.g.
[Bot::Chatbots::Telegram::LongPoll](https://metacpan.org/pod/Bot::Chatbots::Telegram::LongPoll).

These modules rely upon [Log::Any](https://metacpan.org/pod/Log::Any) for emitting logging information; you
are encouraged to read that module's documentation for further
information. For what we are concerned, you have to remember that all logs
will be lost unless you configure [Log::Any](https://metacpan.org/pod/Log::Any), e.g. to send messages on
standard output you can do this:

    use Log::Any::Adapter qw< Stderr >;

If using the web hook, then you will probably want to send the logs
together with [Mojolicious](https://metacpan.org/pod/Mojolicious)'s logs, like this:

    use Mojolicious::Lite;
    use Log::Any::Adapter;
    Log::Any::Adapter->set(MojoLog => logger => app->log);

The configuration above relies on the presence of
[Log::Any::Adapter::MojoLog](https://metacpan.org/pod/Log::Any::Adapter::MojoLog).

# METHODS

All the heavylifting is done by [Bot::ChatBots::MojoPlugin](https://metacpan.org/pod/Bot::ChatBots::MojoPlugin).

# BUGS AND LIMITATIONS

Report bugs through GitHub (patches welcome).

# SEE ALSO

[Bot::ChatBots](https://metacpan.org/pod/Bot::ChatBots), [Bot::ChatBots::Telegram::WebHook](https://metacpan.org/pod/Bot::ChatBots::Telegram::WebHook),
[Bot::ChatBots::Telegram::LongPoll](https://metacpan.org/pod/Bot::ChatBots::Telegram::LongPoll), [WWW::Telegram::BotAPI](https://metacpan.org/pod/WWW::Telegram::BotAPI).

# AUTHOR

Flavio Poletti <polettix@cpan.org>

# COPYRIGHT AND LICENSE

Copyright (C) 2016, 2018 by Flavio Poletti <polettix@cpan.org>

This module is free software. You can redistribute it and/or modify it
under the terms of the Artistic License 2.0.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.


Changes
{{$NEXT}}

0.014  2020-09-11 19:52:08 CEST
   - release after 84 PASSes on CPAN Testers (thanks!)

0.013  2020-09-10 16:24:49 CEST
   - add tracing capabilities to LongPoll
   - add pointers in main module's documentation
   - add logging information in main module's docs
   - upgrade dependency on WWW::Telegram::BotAPI to cope with
     Mojolicious deprecation warning
   - add Test::Pod and Test::Pod::Coverage to modules for
     development

0.012  2018-02-11 15:31:30 CET
   - release after 71 PASSes on CPAN Testers (thanks!)

0.011001  2018-02-11 10:28:52 CET
   - require WWW::Telegram::BotAPI 0.10 or newer
   - move sender accessor in Source role
   - harmonize (send_)?response in WebHook with LongPoll
   - make "sender" available in $record for processors
   - fix error about $log not available during global destruction
   - fix documentation bugs and add enhancements
   - add Tutorial guide and associated examples
   - TRIAL release

0.011  2018-01-27 15:22:09 CET
   - fix chat expansion into channel for CallbackQuery Updates
     (GH#3, thanks xeonvs)
   - document the effect of normalize_record
   - TRIAL release

0.010  2018-01-13 23:28:06 CET
   - remove incorrect assumption about WWW::Telegram::BotAPI
   - release after 71 PASSes on CPAN Testers (thanks!)

0.009001  2018-01-13 22:31:12 CET
   - fix indentation in README.md for examples
   - fix "certificate" and document it in WebHook
   - document constructor in WebHook
   - TRIAL release

0.009  2018-01-11 21:22:52 CET
   - fix minor issue with Kwalitee
   - fix bugs in WebHook.pm and add logging
   - add examples for WebHook and LongPoll

0.008  2018-01-11 08:14:43 CET
   - put "badges" together in POD to avoid indentation effect in
     Markdown
   - remove RT as a channel for tickets
   - update copyright year
   - release after 68 PASSes on CPAN Testers (thanks!)

0.007  2018-01-11 00:40:26 CET
   - fixed issues in documentation
   - add handling of automatic registration/unregistration in
     WebHook
   - fixed bug in LongPoll
   - extended LongPoll for easy send_response
   - TRIAL release

0.006  2017-03-01 22:54:52 CET
   - release after 181 PASSes on CPAN Testers (thanks!)

0.005001  2017-02-27 00:22:46 CET
   - add keyboard id handling

0.005  2017-02-26 19:57:22 CET
   - add handling of keyboards with automatic value translation

0.004  2016-12-06 06:26:28 CET
   - use warnings everywhere

0.002001  2016-11-12 16:54:04 CET
   - add missing dependency on Mock::Quick
   - tracking project via Travis-CI (.travis.yml)
   - badges!

0.002  2016-11-11 23:20:43 CET
   - first release on CPAN (CPAN Testers seems out these days...)


lib/Bot/ChatBots/Telegram/Sender.pm
package Bot::ChatBots::Telegram::Sender;
use strict;
use warnings;
use Ouch;
use 5.010;
{ our $VERSION = '0.013'; }

use WWW::Telegram::BotAPI ();

use Moo;
use namespace::clean;
with 'Bot::ChatBots::Role::Sender';

has start_loop => (
   is => 'rw',
   default => sub { return 0 },
);

has telegram => (
   is      => 'rw',
   lazy    => 1,
   default => sub {
      my $self = shift;
      my $tg   = WWW::Telegram::BotAPI->new(
         token => $self->token,
         async => 1,
      );
      return $tg;
   }
);

has token => (
   is       => 'ro',
   required => 1,
);

# copied from Bot::ChatBot::Role::UserAgent
sub may_start_loop {
   my $self = shift;
   my %args = (@_ && ref($_[0])) ? %{$_[0]} : @_;
   my $start_loop =
     exists($args{start_loop})
     ? $args{start_loop}
     : $self->start_loop;
   Mojo::IOLoop->start if $start_loop && (!Mojo::IOLoop->is_running);
   return $self;
} ## end sub may_start_loop

sub send_message {
   my ($self, $message) = splice @_, 0, 2;
   ouch 500, 'no output to send' unless defined $message;

   my %args = (@_ && ref($_[0])) ? %{$_[0]} : @_;

   # message normalization
   $message =
     ref($message)
     ? {%$message}
     : {text => $message, telegram_method => 'sendMessage'};
   my $method = delete($message->{telegram_method}) // do {
      state $method_for = {
         send        => 'sendMessage',
         sendMessage => 'sendMessage',
      };

      my $name = delete(local $message->{method}) // 'send';
      $method_for->{$name}
        or ouch 500, $self->name . ": unsupported method $name";
   };

   if (($method eq 'sendMessage') && (!exists $message->{chat_id})) {
      if (defined $args{record}) {    # take from $record
         $message->{chat_id} = $args{record}{channel}{id};
      }
      elsif ($self->has_recipient) {
         $message->{chat_id} = $self->recipient;
      }
      else {                          # no more ways to figure it out
         ouch 500, 'no chat identifier for message';
      }
   } ## end if (!exists $message->...)

   my @callback;
   if ($args{callback}) {
      @callback = $args{callback};
   }
   elsif ($self->can('callback')) {
      my $has_callback = $self->can('has_callback') // sub { return 1 };
      @callback = $self->callback if $has_callback->($self);
   }

   my $res = $self->telegram->api_request($method => $message, @callback);

   $self->may_start_loop(%args) if @callback;

   return $res;
} ## end sub send_message

1;


lib/Bot/ChatBots/Telegram/WebHook.pm
package Bot::ChatBots::Telegram::WebHook;
use strict;
use warnings;
{ our $VERSION = '0.013'; }

use Ouch;
use Log::Any qw< $log >;
use Data::Dumper;
use Mojo::URL;
use Mojo::Path;

use Moo;
use namespace::clean;

with 'Bot::ChatBots::Telegram::Role::Source';    # has normalize_record
with 'Bot::ChatBots::Role::WebHook';

has auto_register => (is => 'ro', default => 0, init_arg => 'register');
has auto_unregister => (is => 'rw', default => 0, init_arg => 'unregister');
has certificate => (is => 'rw', default => undef);

sub BUILD {
   my $self = shift;
   $self->install_route;
   $self->register if $self->auto_register;
}

sub DEMOLISH {
   my $self = shift;
   $self->unregister if $self->auto_unregister;
}

sub parse_request {
   my ($self, $req) = @_;
   return $req->json;
}

sub _set_http_response {
   my ($record, $outcome) = @_;

   my $message = $outcome->{sent_response} = {
      method  => 'sendMessage',
      chat_id => $record->{channel}{id},

      ref($outcome->{response}) eq 'HASH'
      ? (%{$outcome->{response}})    # shallow copy suffices
      : (text => $outcome->{response})
   };
   $record->{source}{refs}{controller}->render(json => $message);
   $record->{source}{flags}{rendered} = 1;

   return;
}

around process => sub {
   my ($orig, $self, $record) = @_;

   $record->{source}{refs}{sender} = $self->sender;

   my $outcome = $orig->($self, $record);

   if (ref($outcome) eq 'HASH') {

      # check setting the proper HTTP Response
      # $record and $outcome might be the same, but the flag is
      # namely supported in $record
      _set_http_response($record, $outcome)
         if defined($outcome->{response})
            && (!$record->{source}{flags}{rendered});

      # check using the sender. We use $outcome as $record here, because
      # $outcome is what we are going to pass on eventually
      $self->sender->send_message($outcome->{send_response}, record => $outcome)
         if defined $outcome->{send_response};
   }

   return $outcome;
};

sub register {
   my $self = shift;
   my $args = (@_ && ref($_[0])) ? $_[0] : {@_};

   my $app   = $args->{app}   // $self->app;
   my $token = $args->{token} // $self->token;

   my $wh_url;
   if (my $url = $args->{url} // $self->url) {
      $wh_url = Mojo::URL->new($url);
   }
   else {
      my $path = $args->{path} // $self->path;
      $path = Mojo::Path->new($path);

      my $c = $args->{controller} // $app->build_controller;
      $wh_url = $c->url_for($path);
   } ## end else [ if (my $url = $args->{...})]

   my $wh_url_string = $wh_url->to_abs->to_string;
   my $form = {url => $wh_url_string};

   if (my $certificate = $args->{certificate} // $self->certificate) {
      $certificate = {content => $certificate} unless ref $certificate;
      $form->{certificate} = $certificate;
   }

   $log->info("registering bot URI $wh_url_string");
   $self->_register($args->{token} // $self->token, $form);

   return $self;
} ## end sub register

sub unregister {
   my $self = shift;
   my $args = (@_ && ref($_[0])) ? $_[0] : {@_};
   $self->_register($args->{token} // $self->token);
   return $self;
} ## end sub unregister

sub _register {
   my ($self, $token, $form) = @_;
   require WWW::Telegram::BotAPI;
   my $outcome = WWW::Telegram::BotAPI->new(token => $token)
     ->setWebhook($form // {url => ''});
   $log->info($outcome->{description} // 'unknown result') if $log;
   return;
} ## end sub _register

1;


lib/Bot/ChatBots/Telegram/Keyboard.pm
package Bot::ChatBots::Telegram::Keyboard;
use strict;
use warnings;
{ our $VERSION = '0.013'; }

use Ouch;
use Log::Any qw< $log >;
use Data::Dumper;

use Moo;
use namespace::clean;

use Exporter qw< import >;
our @EXPORT_OK = qw< keyboard >;

has displayable => (
   is       => 'ro',
   required => 1,
);

has id => (
   is      => 'ro',
   default => sub { return 0 },
   isa     => sub {
      my $n         = shift;
      my $complaint = 'keyboard_id MUST be an unsigned 32 bits integer';
      ouch 500, $complaint unless $n =~ m{\A(?: 0 | [1-9]\d* )\z}mxs;
      my $r = unpack 'N', pack 'N', $n;
      ouch 500, $complaint unless $n eq $r;
      return;
   },
);

has _value_for => (
   is       => 'ro',
   required => 1,
);

{
   my ($ONE, $ZERO, $BOUNDARY);

   BEGIN {
      $ONE      = "\x{200B}";
      $ZERO     = "\x{200C}";
      $BOUNDARY = "\x{200D}";
   } ## end BEGIN

   sub __encode_uint32 {
      my $x = shift;
      (my $b = unpack 'B32', pack 'N', $x) =~ s/^0+//mxs;
      $b = '0' unless length $b;
      return join '', map { $_ ? $ONE : $ZERO } split //, $b;
   } ## end sub __encode_uint32

   sub __decode_uint32 {
      my $x = shift;
      my $b = join '', map { $_ eq $ONE ? '1' : '0' } split //, $x;
      $b = substr(('0' x 32) . $b, -32, 32);
      return unpack 'N', pack 'B32', $b;
   } ## end sub __decode_uint32

   sub __encode {
      my ($label, $keyboard_id, $code) = @_;
      return join '', $label,
        $BOUNDARY, __encode_uint32($keyboard_id),
        $BOUNDARY, __encode_uint32($code),
        $BOUNDARY;
   } ## end sub __encode

   sub __decode {
      return unless defined $_[0];
      my ($label, $kid, $code) = $_[0] =~ m{
         \A
                      (.*)
            $BOUNDARY ((?:$ZERO|$ONE)+)
            $BOUNDARY ((?:$ZERO|$ONE)+)
            $BOUNDARY
         \z
      }mxs;
      return unless defined $code;
      return ($label, __decode_uint32($kid), __decode_uint32($code));
   } ## end sub __decode
}

sub BUILDARGS {
   my ($class, %args) = @_;
   ouch 500, 'no input keyboard' unless exists $args{keyboard};
   my $id = $args{id} //= 0;
   @args{qw<displayable _value_for>} = __keyboard($args{keyboard}, $id);
   return \%args;
} ## end sub BUILDARGS

sub _decode {
   my ($self, $x, $name) = @_;
   if (ref($x) eq 'HASH') {
      $x = $x->{payload} if exists $x->{payload};
      $x = $x->{text} // undef;
   }
   elsif (ref($x)) {
      ouch 500, "$name(): pass either hash references or plain scalars";
   }

   return __decode($x);
} ## end sub _decode

sub get_value {
   my ($self, $x) = @_;
   my (undef, undef, $code) = $self->_decode($x, 'get_value');
   return undef unless defined $code;

   my $vf = $self->_value_for;
   if (!exists($vf->{$code})) {
      $log->warn("get_value(): received code $code is unknown");
      return undef;
   }
   return $vf->{$code};
} ## end sub get_value

sub get_keyboard_id {
   my ($self, $x) = @_;
   my (undef, $keyboard_id) = $self->_decode($x, 'get_keyboard_id');
   return $keyboard_id;
}

sub __keyboard {
   my ($input, $keyboard_id) = @_;
   ouch 500, 'invalid input keyboard, not an ARRAY'
     unless ref($input) eq 'ARRAY';
   ouch 500, 'invalid empty keyboard' unless @$input;

   my $code = 0;
   my @display_keyboard;
   my (%value_for, %code_for);
   for my $row (@$input) {
      ouch 500, 'invalid input keyboard, not an AoA'
        unless ref($row) eq 'ARRAY';

      my @display_row;
      push @display_keyboard, \@display_row;
      for my $item (@$row) {
         ouch 500, 'invalid input keyboard, not an AoAoH'
           unless ref($item) eq 'HASH';

         my %display_item = %$item;
         push @display_row, \%display_item;

         my $command = delete $display_item{_value};
         next unless defined $command;
         my $cc = $code_for{$command} //= $code++;
         $value_for{$cc} //= $command;
         $display_item{text} =
           __encode($display_item{text}, $keyboard_id, $cc);
      } ## end for my $item (@$row)
   } ## end for my $row (@$input)
   return (\@display_keyboard, \%value_for);
} ## end sub __keyboard

sub keyboard {
   my %args;
   if (@_ > 1) {
      if (ref($_[0])) {
         $args{keyboard} = [@_];
      }
      else {
         %args = @_;
      }
   } ## end if (@_ > 1)
   elsif (@_ == 1) {
      my $x = shift;
      if (@$x > 0) {
         if (ref($x->[0]) eq 'ARRAY') {
            $args{keyboard} = $x;
         }
         else {
            $args{keyboard} = [$x];    # one row only
         }
      } ## end if (@$x > 0)
   } ## end elsif (@_ == 1)
   return Bot::ChatBots::Telegram::Keyboard->new(%args);
} ## end sub keyboard

1;


lib/Bot/ChatBots/Telegram/LongPoll.pm
package Bot::ChatBots::Telegram::LongPoll;
use strict;
use warnings;
{ our $VERSION = '0.013'; }

use Ouch;
use Try::Tiny;
use Log::Any qw< $log >;
use Mojo::IOLoop ();
use IO::Socket::SSL ();    # just to be sure to complain loudly in case
use List::Util qw< max >;
use Data::Dumper;

use Moo;
use namespace::clean;

with 'Bot::ChatBots::Telegram::Role::Source';    # normalize_record, token
with 'Bot::ChatBots::Role::Source';              # processor, typename

has connect_timeout => (
   is      => 'ro',
   default => sub { return 20 },
);

has interval => (
   is      => 'ro',
   default => sub { return 0.1 },
);

has max_redirects => (
   is => 'ro',
   default => sub { return 5 },
);

has _start => (
   is       => 'ro',
   default  => sub { return 1 },
   init_arg => 'start',
);

has update_timeout => (
   is      => 'ro',
   default => sub { return 300 },
);

sub BUILD {
   my $self = shift;
   $self->start if $self->_start;
}

sub class_custom_pairs {
   my $self = shift;
   return (token => $self->token);
}

sub parse_response {
   my ($self, $res, $threshold_id) = @_;
   my $data = $res->json // {};
   return grep { $_->{update_id} >= $threshold_id } @{$data->{result}//[]}
      if $data->{ok}; # boolean flag from Telegram API

   my $error = $data->{description} // 'unknown error';
   $log->error('getUpdates error: ' . $error);
   if ($log->is_trace) {
      local $Data::Dumper::Indent = 1;
      for ([json => $data], [res => $res]) {
         (my $d = Dumper $_->[1]) =~ s{\A.*?=}{$_->[0] =>}mxs;
         $log->trace($d);
      }
   }
   return;
}

sub poller {
   my $self = shift;
   my $args = (@_ && ref($_[0])) ? $_[0] : {@_};

   my $update_timeout = $self->update_timeout;
   my %query = (
      offset => 0,
      telegram_method => 'getUpdates',
      timeout => $update_timeout,
   );

   my $sender = $self->sender;
   $sender->telegram->agent->connect_timeout($self->connect_timeout)
     ->inactivity_timeout($update_timeout + 5)
     ->max_redirects($self->max_redirects);

   # this flag tells us whether we're in a call already, avoiding
   # duplicates. It is set before sending a request, and reset when the
   # response is managed
   my $is_busy;

   my $on_data = sub {
      my ($ua, $tx) = @_;

      my @updates;
      try {
         @updates = $self->parse_response($tx->res, $query{offset});
      }
      catch {
         $log->error(bleep $_);
         die $_ if $self->should_rethrow($args);
      };

      my @retval = $self->process_updates(
         refs => {
            sender => $sender,
            tx     => $tx,
            ua     => $ua,
         },
         source_pairs => {
            query => \%query,
         },
         updates => \@updates,
         %$args, # may override it all!
      );

      for my $item (@retval) {
         next unless defined $item;
         defined(my $record = $item->{record})            or next;
         defined(my $outcome = $item->{outcome})          or next;
         defined(my $message = $outcome->{send_response}) or next;
         $sender->send_message($message, record => $record);
      }

      # if we get here, somehow me managed to get past this call... Get
      # ready for the next one. Just to be on the safe side, we will
      # advance $query{offset} anyway
      $query{offset} = 1 + max map { $_->{update_id} } @updates
         if @updates;
      $is_busy = 0;
   };

   return sub {
      return if $is_busy;
      $is_busy = 1; # $on_data below will reset $is_busy when ready
      $sender->send_message(\%query, callback => $on_data);
   };
} ## end sub callback

around process => sub {
   my ($orig, $self, $record) = @_;
   my $outcome = $orig->($self, $record);
   $record->{source}{query}{offset} = $record->{update}{update_id} + 1;
   return $outcome;
};

sub start {
   my $self = shift;
   Mojo::IOLoop->recurring($self->interval, $self->poller(@_));
   Mojo::IOLoop->start unless Mojo::IOLoop->is_running;
   return $self;
}

1;


lib/Bot/ChatBots/Telegram/Role/Source.pm
package Bot::ChatBots::Telegram::Role::Source;
use strict;
use warnings;
{ our $VERSION = '0.013'; }

use Ouch;
use Log::Any;

use Moo::Role;

has token => (
   is => 'ro',
   lazy => 1,
   predicate => 1,
   default => sub { ouch 500, 'token is not defined' }
);

has sender => (
   is      => 'ro',
   lazy    => 1,
   default => sub {    # prefer has-a in this case
      my $self = shift;
      require Bot::ChatBots::Telegram::Sender;
      return Bot::ChatBots::Telegram::Sender->new(token => $self->token);
   },
);

{
   my %data_type_for = (
      message => 'Message',
      edited_message => 'Message',
      channel_post => 'Message',
      edited_channel_post => 'Message',
      inline_query => 'InlineQuery',
      chosen_inline_result => 'ChosenInlineResult',
      callback_query => 'CallbackQuery',
      shipping_query => 'ShippingQuery',
      pre_checkout_query => 'PreCheckoutQuery',
   );

   sub normalize_record {
      my ($self, $record) = @_;

      my $update = $record->{update} or ouch 500, 'no update found!';
      $record->{source}{technology} = 'telegram';
      $record->{source}{token} //= $record->{source}{object_token};

      my ($type) = grep { $_ ne 'update_id' } keys %$update;
      $record->{type} = $type;

      $record->{data_type} = $data_type_for{$type} || 'unknown';

      my $payload = $record->{payload} = $update->{$type};

      $record->{sender} = $payload->{from};

      return $self->_normalize_record_chan($record);
   }
}

sub _normalize_record_chan {
   my ($self, $record) = @_;
   my ($dtype, $payload) = @{$record}{qw< data_type payload >};
   my $chan;
   if ($dtype eq 'Message') {
      $chan = {%{$payload->{chat}}};
   }
   elsif ($dtype eq 'CallbackQuery') {
      if (exists $payload->{message}) {
         $chan = {%{$payload->{message}{chat}}};
      }
      else { # FIXME guessing correctly here?
         $chan = {id => $payload->{chat_instance}};
      }
   }
   if ($chan) {
      $chan->{fqid} = "$chan->{type}/$chan->{id}" if exists $chan->{id};
      $record->{channel} = $chan;
   }

   return $record;
}

1;
